docs/sphinx/examples/task_based_programming.rst

```diff
@@ -0,0 +1,140 @@
+..
+    Copyright (C) 2019 Tapasweni Pathak
+
+    Distributed under the Boost Software License, Version 1.0. (See accompanying
+    file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+.. _examples_task_based_programming:
+
+================================================================
+Designing and Writie task Based Programs: Task Based Programming
+================================================================
+
+Task based programming is a different approach to writing your code. In task
+based programming
+
+- your program should be a tree of tasks
+- ach task will depend on other tasks
+- and have child tasks
+- you don't have to have the same on all nodes (c.f. MPI)
+
+.. figure:: /_static/images/DAG-weather.jpg
+
+- each one of these boxes will contain a sub-tree of (possibly millions) of
+  tasks
+
+Task Goals
+=====
+
+- Make these gaps as small as possible
+  - Keep breaking tasks into smaller tasks so the scheduler can fill gaps
+  - limit of task size/granularity is a function of overheads
+    - context switch (actually swapping stack, registers etc)
+    - time to create, queue, dequeue a task
+
+
+.. figure:: /_static/images/Task-waits.jpg
+
+Bad tasks (or scheduling)
+=========================
+
+.. figure:: /_static/images/tasks-bad.jpg
+
+Good tasks
+==========
+
+.. figure:: /_static/images/tasks-good.jpg
+
+Task decomposition
+==================
+
+- Breaking a program into tasks should be straightforward
+- More functional
+  - tasks should accept inputs and return results
+  - modifying global state should be avoided
+    - race conditions and other thread related problems (deadlocks etc)
+  - the leaf nodes of the tree are the smallest bits of work you can express
+    - but those leaf nodes might be broken further by HPX
+    - even ``parallel::for(...)`` loops decompose into tasks
```

Comment:
`parallel::for()` is not a thing, it might be better to either say `parallel:::for_each()` or simply refer to all parallel algorithms